### ISSUE TITLE
LemMinX loading local artifacts doesn't seem to finish and makes whole of Eclipse a bitlaggy

### ISSUE THREAD

**Issue by mbrandl87** [eclipse/lemminx-maven#517]
Hi there,

since I updated Eclipse to 23-09 this task seems to run all the time and never getting through.
In Progress it states LemMinX - loading local artifacts from <my local maven repo folder> , 170k of 200k... 
Which in turn is strange as well, the folder doesn't even have 200k  files :)
It started shortly after I first opened a pom.xml. When I interrupt it from progress view, it comes up again in the next pom.

This constantly consumes ~10% of CPU. Its Running since 5 hours now and the numbers don't move.

Is there anything I can check or some way to speed it up? :)

Best
Martin

**Comment by angelozerr** [comment]
>  When I interrupt it from progress view, it comes up again in the next pom.

Interrupt doesn't interrupt the job, see issue https://github.com/eclipse/lsp4e/issues/742

> In Progress it states LemMinX - loading local artifacts from , 170k of 200k...
>Which in turn is strange as well, the folder doesn't even have 200k files :)

Have  you this problem with another local repo?

**Comment by mbrandl87** [comment]
I'm only using one maven local repo folder so far, lets see what happens when I try a fresh one.

**Comment by mbrandl87** [comment]
It looks like that somehow did the trick. I Just opened the pom again and this time the scan showed up for some seconds and then disappeared. But is this just a matter of size? The original had about 57k files with ~15 gb. The new one is just at ~2 gb sofar.
I noted that this thime the .lemminx-maven also contains a part of the repository - why does it download the artifacts to two places now? With the old repo this folder just contained lifecycle-mapping-1.0.0.jar.lastUpdated with the corresponding parrent folder structure as if it got stuck there.

**Comment by angelozerr** [comment]
> I noted that this thime the .lemminx-maven also contains a part of the repository - why does it download the artifacts to two places now?

It is to avoid polluating your real local repository when you are typing, see https://github.com/eclipse/lemminx-maven/issues/213#issuecomment-1669298790

**Comment by mbrandl87** [comment]
Ah I see. Seems fine now. Do you have an idea how I could track down what in the repo caused the problem? Otherwise I have to keep a copy of the old one, because some artifacts can't be retrieved automatically., e.g. eclipse factory path entries.

**Comment by markuswernig** [comment]
Sorry for reviving this thread ...

The issue still persists and it makes Eclipse almost unusable now... The org.eclipse.lemminx.XMLServerLauncher process that gets started when a pom.xml is opened, uses 100% of one CPU, and it never stops. The fan on my laptop is spinning at max all the time.

I did an strace -f -p {pid of XMLServerLauncher process}, and found that is is constantly scanning my home directory. Scanning means opening every single file with statx(...), then closing it. Doing so, it appears it blindly follows symlinks, regardless of where they point. If they point to directories, those directories get scanned as well.

In the case where symlinks point out of my home directory (more precisely to "/"), that leads to a circular read loop.
Imagine this layout

/
/var/
/usr/
/home/
/home/user
/home/user/a_symlink_to_root -> /
/home/user/regular_file

Now XMLServerLauncher scans /home/user, finds /home/user/a_symlink_to_root and starts scanning /, until it finds /home/user again and in it /home/user/a_symlink_to_root, starts scanning / again, and so on and so forth in all eternity. 
It actually never reaches /home/user/regular_file, btw.

The progress bar stays at LemMinX: (90%), with the text "Loading local artifacts from ""...: Scanning folder ".mypaint" (1915 / 2127)...

If I close the pom, the XMLServerLauncher process eventually stops (after some minutes). 

The next time I open a pom again, it all starts over, againg hanging at 90%.
If I remove the ".mypaint" folder, it will display another folder name and other numbers, still hanging at 90%.

The OS (Linux in this case) does eventually (after ~10 min) cancel the statx() calls with "Too many levels of symbolic links" errors, but that does not stop the XMLServerLauncher from continuing. It seems to just go up one level and continue from there, always running into the same dead end.

Not sure how to implement this, but I think there needs to be a protection against circular reads.

**Comment by fbricon** [comment]
can you report the result of `jstack <lemminx-PID>`?

**Comment by markuswernig** [comment]
The list of "at org.eclipse.lemminx.extensions.maven.searcher.LocalRepositorySearcher$Cache.updateArtifacts(LocalRepositorySearcher.java:241)" lines under "ForkJoinPool.commonPool-worker-1" https://github.com/eclipse/lemminx-maven/pull/31 [10685]
gets longer with every invocation of jstack <pid>

**Comment by markuswernig** [comment]
Definitely.

But there's also the question why lemminx would look for artifacts in my entire home directory in the first place. 
As I have the habit of keeping old artifacts (and old versions of my .m2 repo) in my home directory, this could lead to some funny behaviour, don't you think?

**Comment by fbricon** [comment]
lemminx-maven indexes your local repository so it can provide content-assist when editing `<dependency>` nodes in your pom.xml. I don't think it exposes any preference to disable that behavior (which would need to be surfaced in Eclipse/m2e as well).

**Comment by markuswernig** [comment]
Yes, agreed. But what is the definition of "local repository"? I'd expect it to index the repositories under ~/.m2, but not everything in my home directory. Can the scope of "local repository" be configured anywhere in Eclipse?
